cypress/integration/pages.spec.js

```diff
@@ -0,0 +1,639 @@
+      cy.contains('My Workspace').click()
+      cy.contains(PRETTIFIED_PAGE_TITLE)
+    })
+
+    it('Should not be able to create page with invalid title or permalink', () => {
+      const SHORT_TITLE = 'abc'
+      const SHORT_PERMALINK = '/12'
+      const INVALID_PERMALINK = 'test-'
+
+      cy.get('#settings-NEW').click()
+
+      // Page title has to be more than 4 characters long
+      cy.get('#title').clear().type(SHORT_TITLE)
+      cy.get('#permalink').clear().type(TEST_PAGE_PERMALNK)
+      cy.contains('The title should be longer than 4 characters.')
+      cy.contains('button', 'Save').should('be.disabled')
+  
+      // Page title must not already exist
+      cy.get('#title').clear().type(TEST_PAGE_TITLE)
+      cy.contains('This title is already in use. Please choose a different title.')
+      cy.contains('button', 'Save').should('be.disabled')
+
+      // Permalink needs to be longer than 4 characters
+      cy.get('#title').clear().type(EDITED_TEST_PAGE_TITLE)
+      cy.get('#permalink').clear().type(SHORT_PERMALINK)
+      cy.contains('The permalink should be longer than 4 characters.')
+      cy.contains('button', 'Save').should('be.disabled')
+  
+      // Permalink should start with a slash, and contain alphanumeric characters separated by hyphens and slashes only
+      cy.get('#title').clear().type(EDITED_TEST_PAGE_TITLE)
+      cy.get('#permalink').clear().type(INVALID_PERMALINK)
+      cy.contains('The url should start with a slash, and contain alphanumeric characters separated by hyphens and slashes only.')
+      cy.contains('button', 'Save').should('be.disabled')
+    })
+
+    it('Should be able to edit existing page details with valid title and permalink', () => {
+      const testPageCard = cy.contains(PRETTIFIED_PAGE_TITLE)
+
+      // User should be able edit page details
+      testPageCard.children().within(() => cy.get('[id^=settings-]').click())
+      cy.get('div[id^=settings-]').first().click() // .first() is necessary because the get returns multiple elements (refer to MenuDropdown.jsx)
+      cy.get('#title').clear().type(EDITED_TEST_PAGE_TITLE)
+      cy.contains('button', 'Save').click()
+
+      // New page title should be reflected in the Workspace
+      const editedTestPageCard = cy.contains(EDITED_PRETTIFIED_PAGE_TITLE)
+
+      // Reset the page title to previous title
+      editedTestPageCard.children().within(() => cy.get('[id^=settings-]').click())
+      cy.get('div[id^=settings-]').first().click() // .first() is necessary because the get returns multiple elements (refer to MenuDropdown.jsx)
+      cy.get('#title').clear().type(TEST_PAGE_TITLE)
+      cy.contains('button', 'Save').click()
+      cy.wait(4000)
+
+      // Page title should be reset in the Workspace
+      cy.contains(PRETTIFIED_PAGE_TITLE)
+      cy.wait(4000) // Wait needed because the next test fails otherwise - the rename call takes some time to propagate
```

Comment:
If we're using `cy.wait()`  pretty often, should we set the default timeout to be a larger value? Part of the appeal of using Cypress is to avoid having to code for these edge cases (reference: [Cypress docs](https://docs.cypress.io/guides/core-concepts/introduction-to-cypress#Cypress-is-Not-Like-jQuery)]). Let me know if I'm missing something

Reply:
@kwajiehao I read a little more about using `cy.wait()` and it turns out to not be good practice. This is because as you've pointed out, if the wait time is too short, the tests are likely to flake. But on the other hand, if the wait time is too long, the time to run the entire test suite increases significantly.

Instead, I've made a change in 
b8fae3d to do the equivalent of wait by using `cy.should('exist')` with a very long timeout. `.should()` will keep retrying until it hits the timeout or until the assertion is met ([docs](https://docs.cypress.io/api/commands/should#Timeouts)). This way, if the awaited step completes early, the test will move on.